File: js/core/transforms.js
// Coordinate-frame transforms and conversions.
//
// Reference frame: stationary flat earth, rotating celestial dome.
//
//   celest  - +z is the celestial pole (dome's axis of rotation);
//             sun/moon directions live here as unit vectors
//   globe   - local observer frame: +x zenith, +y east, +z north
//   fe-local- local flat-earth frame at observer: +z up, +x outward, +y east
//   fe      - global stationary flat-earth disc: z=0 is the disc plane, +z up
//   dome    - rotating sky frame: sky's current angular position about +z
//             determines how a dome-fixed point maps to fe coords

import { ToRad, ToDeg, Limit1 } from '../math/utils.js';
import { M } from '../math/mat3.js';
import { V } from '../math/vect3.js';
import { canonicalLatLongToDisc } from './canonical.js';

// Celest -> local-globe: rotate by -(observer-longitude + skyRotAngle) about Z,
// then by +latitude about Y.
export function compTransMatCelestToGlobe(obsLatDeg, obsLongDeg, skyRotAngleDeg) {
  const first = M.RotatingZ(ToRad(-obsLongDeg - skyRotAngleDeg));
  return M.RotatingY(ToRad(obsLatDeg), first);
}

// Local-fe -> global-fe.
//
// Local-FE convention: +x = south, +y = east, +z = up. The rotation
// is built from the active projection's gradient at (obsLat, obsLon)
// so a non-AE projection (currently only DP) gets the right
// curved-meridian tangent for "north". Sampling
// `canonicalLatLongToDisc` keeps this in lockstep with the
// projection-dispatch in canonical.js (S680).
//
// For AE polar, the gradient collapses to `-(cos λ, sin λ)` for the
// "north" direction — i.e. `RotatingZ(λ)` — so behaviour is unchanged
// when no override is active. At the canonical AE pole (lat = 90°)
// the gradient is singular; fall back to the longitude rotation so
// the meridian picked by ObserverLong continues to drive the frame.
export function compTransMatLocalFeToGlobalFe(observerCoord, observerLongDeg, observerLatDeg = null) {
  let ax, ay, bx, by;
  if (observerLatDeg === null) {
    const lr = ToRad(observerLongDeg);
    const cl = Math.cos(lr), sl = Math.sin(lr);
    ax = cl;  ay = sl;       // local +x (south) → global
    bx = -sl; by = cl;       // local +y (east)  → global
  } else {
    const eps = 1e-3;
    const lat = observerLatDeg;
    const lon = observerLongDeg;
    const pHere = canonicalLatLongToDisc(lat, lon, 1);
    const latProbe = lat >= 90 - eps ? lat - eps : lat + eps;
    const sign = lat >= 90 - eps ? -1 : 1;
    const pN = canonicalLatLongToDisc(latProbe, lon, 1);
    const dnx = (pN[0] - pHere[0]) * sign;
    const dny = (pN[1] - pHere[1]) * sign;
    const nLen = Math.hypot(dnx, dny);
    if (nLen < 1e-9) {
      const lr = ToRad(observerLongDeg);
      const cl = Math.cos(lr), sl = Math.sin(lr);
      ax = cl;  ay = sl;
      bx = -sl; by = cl;
    } else {
      const nx = dnx / nLen, ny = dny / nLen;
      ax = -nx; ay = -ny;       // south = -north
      bx =  ny; by = -nx;       // east  = north rotated −90° in disc plane
    }
  }
  const rot = {
    r: [
      [ax, bx, 0],
      [ay, by, 0],
      [0,  0,  1],
    ],
    t: [0, 0, 0],
  };
  return M.Moving(observerCoord[0], observerCoord[1], observerCoord[2], rot);
}

// Dome -> fe: the sky-frame point, after the current sky rotation, expressed
// in the stationary disc's coordinates.
export function compTransMatVaultToFe(skyRotAngleDeg) {
  return M.RotatingZ(-ToRad(skyRotAngleDeg));
}

// --- Point conversions ----------------------------------------------------

export function celestCoordToLocalGlobeCoord(celestCoord, transMatCelestToGlobe) {
  return M.Trans(transMatCelestToGlobe, celestCoord);
}

// Spherical (long, lat, r) -> cartesian.
export function latLongToCoord(latDeg, longDeg, length) {
  return V.FromAngle(longDeg, latDeg, length);
}

// Cartesian -> { lat, lng } in degrees.
export function coordToLatLong(coord) {
  const vectXY = [coord[0], coord[1], 0];
  const xyLen = V.Length(vectXY);
  if (xyLen === 0) {
    return { lat: coord[2] >= 0 ? 90 : -90, lng: 0 };
  }
  const xyNorm = V.Norm(vectXY);
  const norm   = V.Norm(coord);
  const lat = 90 - ToDeg(Math.acos(Limit1(V.ScalarProd([0, 0, 1], norm))));
  let   lng = ToDeg(Math.acos(Limit1(V.ScalarProd([1, 0, 0], xyNorm))));
  if (xyNorm[1] < 0) lng *= -1;
  return { lat, lng };
}

// Equatorial (RA / Dec) → horizontal (az / el) at the observer's
// (lat, lon) and the supplied Greenwich mean sidereal time.
// RA / Dec in radians; lat / lon / gmst in degrees. Returns
// `{ azimuth, elevation }` in degrees, with azimuth measured
// clockwise from north and elevation 0° at the horizon. Returns
// NaN/NaN when the input RA / Dec aren't finite — handy when a
// comparison pipeline reports NaN for an unsupported body.
export function raDecToAzEl(raRad, decRad, latDeg, lonDeg, gmstDeg) {
  if (!Number.isFinite(raRad) || !Number.isFinite(decRad)) {
    return { azimuth: NaN, elevation: NaN };
  }
  const DEG = Math.PI / 180;
  const lat = latDeg * DEG;
  const lst = (gmstDeg + lonDeg) * DEG;
  const ha  = lst - raRad;
  const sinAlt = Math.sin(lat) * Math.sin(decRad)
              + Math.cos(lat) * Math.cos(decRad) * Math.cos(ha);
  const alt = Math.asin(Math.max(-1, Math.min(1, sinAlt)));
  const y = -Math.cos(decRad) * Math.sin(ha);
  const x = Math.sin(decRad) * Math.cos(lat)
          - Math.cos(decRad) * Math.sin(lat) * Math.cos(ha);
  let az = Math.atan2(y, x) * 180 / Math.PI;
  az = ((az % 360) + 360) % 360;
  return { azimuth: az, elevation: alt * 180 / Math.PI };
}

// Local-globe cartesian -> { azimuth, elevation } in degrees.
// Convention: x=zenith, y=east, z=north.
export function localGlobeCoordToAngles(coord) {
  const yzLen = Math.hypot(coord[1], coord[2]);
  const norm = V.Norm(coord);
  let azimuth;
  if (yzLen === 0) {
    azimuth = 0;
  } else {
    const yzNorm = [0, coord[1] / yzLen, coord[2] / yzLen];
    azimuth = ToDeg(Math.acos(Limit1(V.ScalarProd([0, 0, 1], yzNorm))));
    if (yzNorm[1] < 0) azimuth = 360 - azimuth;
  }
  const elevation = 90 - ToDeg(Math.acos(Limit1(V.ScalarProd([1, 0, 0], norm))));
  return { azimuth, elevation };
}

// Swap of axis convention from local-globe (x-zenith, y-east, z-north) to
// local-fe (x-north, y-east, z-up).
export function localGlobeCoordToLocalFeCoord(v) {
  return [-v[2], v[1], v[0]];
}

export function localGlobeCoordToGlobalFeCoord(v, transMatLocalFeToGlobalFe) {
  return M.Trans(transMatLocalFeToGlobalFe, localGlobeCoordToLocalFeCoord(v));
}

export function vaultCoordToGlobalFeCoord(v, transMatDomeToFe) {
  return M.Trans(transMatDomeToFe, v);
}

// FE-conceptual unit direction in the observer's local-globe frame
// (x = zenith, y = east, z = north).
//
// Given a body's vault position in global FE coords and the
// observer's global FE position, this returns the unit ray
// observer → body expressed in local-globe axes — i.e. what the
// observer sees on a flat-earth conceptual model where the active
// projection (currently DP) places both the observer and the vault
// position on the disc. Drop-in replacement for
// `celestCoordToLocalGlobeCoord(...)` when you want the optical
// vault to follow the FE projection rather than the sphere model.
export function feConceptualLocalGlobeUnit(vaultGlobalFe, observerGlobalFe, transMatLocalFeToGlobalFe) {
  const dx = vaultGlobalFe[0] - observerGlobalFe[0];
  const dy = vaultGlobalFe[1] - observerGlobalFe[1];
  const dz = vaultGlobalFe[2] - observerGlobalFe[2];
  // Inverse rotation: rotation matrix is orthonormal so transpose suffices.
  const r = transMatLocalFeToGlobalFe.r;
  const lfX = r[0][0] * dx + r[1][0] * dy + r[2][0] * dz; // local-FE x = south
  const lfY = r[0][1] * dx + r[1][1] * dy + r[2][1] * dz; // local-FE y = east
  const lfZ = r[0][2] * dx + r[1][2] * dy + r[2][2] * dz; // local-FE z = up
  // local-FE (south, east, up) → local-globe (zenith, east, north) = (z, y, -x)
  const lgZenith = lfZ;
  const lgEast   = lfY;
  const lgNorth  = -lfX;
  const len = Math.hypot(lgZenith, lgEast, lgNorth);
  if (len < 1e-12) return [0, 0, 0];
  return [lgZenith / len, lgEast / len, lgNorth / len];
}

File: src/transforms.rs
use crate::math::{limit1, to_deg, to_rad};
use crate::matrix::Transform3;
use crate::projections::{canonical_lat_long_to_disc, ProjectionId};
use crate::vector::Vec3;

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct LatLong {
    pub lat: f64,
    pub lng: f64,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Angles {
    pub azimuth: f64,
    pub elevation: f64,
}

pub fn comp_trans_mat_celest_to_globe(
    obs_lat_deg: f64,
    obs_lon_deg: f64,
    sky_rot_angle_deg: f64,
) -> Transform3 {
    let first = Transform3::rotating_z(to_rad(-obs_lon_deg - sky_rot_angle_deg), None);
    Transform3::rotating_y(to_rad(obs_lat_deg), Some(first))
}

pub fn comp_trans_mat_local_fe_to_global_fe(
    observer_coord: Vec3,
    observer_lon_deg: f64,
    observer_lat_deg: Option<f64>,
    active_projection: Option<ProjectionId>,
) -> Transform3 {
    let (ax, ay, bx, by) = if let Some(observer_lat_deg) = observer_lat_deg {
        let eps = 1e-3;
        let lat = observer_lat_deg;
        let lon = observer_lon_deg;
        let p_here = canonical_lat_long_to_disc(lat, lon, 1.0, active_projection);
        let lat_probe = if lat >= 90.0 - eps { lat - eps } else { lat + eps };
        let sign = if lat >= 90.0 - eps { -1.0 } else { 1.0 };
        let p_n = canonical_lat_long_to_disc(lat_probe, lon, 1.0, active_projection);
        let dnx = (p_n.x - p_here.x) * sign;
        let dny = (p_n.y - p_here.y) * sign;
        let n_len = (dnx * dnx + dny * dny).sqrt();
        if n_len < 1e-9 {
            simple_local_axes(observer_lon_deg)
        } else {
            let nx = dnx / n_len;
            let ny = dny / n_len;
            (-nx, -ny, ny, -nx)
        }
    } else {
        simple_local_axes(observer_lon_deg)
    };

    let rot = Transform3 {
        r: [[ax, bx, 0.0], [ay, by, 0.0], [0.0, 0.0, 1.0]],
        t: Vec3::ZERO,
    };
    Transform3::moving(observer_coord.x, observer_coord.y, observer_coord.z, Some(rot))
}

fn simple_local_axes(observer_lon_deg: f64) -> (f64, f64, f64, f64) {
    let lr = to_rad(observer_lon_deg);
    let cl = lr.cos();
    let sl = lr.sin();
    (cl, sl, -sl, cl)
}

pub fn comp_trans_mat_vault_to_fe(sky_rot_angle_deg: f64) -> Transform3 {
    Transform3::rotating_z(-to_rad(sky_rot_angle_deg), None)
}

pub fn celest_coord_to_local_globe_coord(
    celest_coord: Vec3,
    trans_mat_celest_to_globe: Transform3,
) -> Vec3 {
    trans_mat_celest_to_globe.trans(celest_coord)
}

pub fn lat_long_to_coord(lat_deg: f64, lon_deg: f64, length: f64) -> Vec3 {
    Vec3::from_angle(lon_deg, lat_deg, length)
}

pub fn coord_to_lat_long(coord: Vec3) -> LatLong {
    let vect_xy = Vec3::new(coord.x, coord.y, 0.0);
    let xy_len = vect_xy.length();
    if xy_len == 0.0 {
        return LatLong {
            lat: if coord.z >= 0.0 { 90.0 } else { -90.0 },
            lng: 0.0,
        };
    }
    let xy_norm = vect_xy.norm();
    let norm = coord.norm();
    let lat = 90.0 - to_deg(limit1(Vec3::new(0.0, 0.0, 1.0).dot(norm)).acos());
    let mut lng = to_deg(limit1(Vec3::new(1.0, 0.0, 0.0).dot(xy_norm)).acos());
    if xy_norm.y < 0.0 {
        lng *= -1.0;
    }
    LatLong { lat, lng }
}

pub fn local_globe_coord_to_angles(coord: Vec3) -> Angles {
    let yz_len = (coord.y * coord.y + coord.z * coord.z).sqrt();
    let norm = coord.norm();
    let azimuth = if yz_len == 0.0 {
        0.0
    } else {
        let yz_norm = Vec3::new(0.0, coord.y / yz_len, coord.z / yz_len);
        let mut az = to_deg(limit1(Vec3::new(0.0, 0.0, 1.0).dot(yz_norm)).acos());
        if yz_norm.y < 0.0 {
            az = 360.0 - az;
        }
        az
    };
    let elevation = 90.0 - to_deg(limit1(Vec3::new(1.0, 0.0, 0.0).dot(norm)).acos());
    Angles { azimuth, elevation }
}

pub fn local_globe_coord_to_local_fe_coord(v: Vec3) -> Vec3 {
    Vec3::new(-v.z, v.y, v.x)
}

pub fn local_globe_coord_to_global_fe_coord(
    v: Vec3,
    trans_mat_local_fe_to_global_fe: Transform3,
) -> Vec3 {
    trans_mat_local_fe_to_global_fe.trans(local_globe_coord_to_local_fe_coord(v))
}

pub fn vault_coord_to_global_fe_coord(v: Vec3, trans_mat_vault_to_fe: Transform3) -> Vec3 {
    trans_mat_vault_to_fe.trans(v)
}

pub fn fe_conceptual_local_globe_unit(
    vault_global_fe: Vec3,
    observer_global_fe: Vec3,
    trans_mat_local_fe_to_global_fe: Transform3,
) -> Vec3 {
    let d = vault_global_fe.sub(observer_global_fe);
    let r = trans_mat_local_fe_to_global_fe.r;
    let lf_x = r[0][0] * d.x + r[1][0] * d.y + r[2][0] * d.z;
    let lf_y = r[0][1] * d.x + r[1][1] * d.y + r[2][1] * d.z;
    let lf_z = r[0][2] * d.x + r[1][2] * d.y + r[2][2] * d.z;
    let local_globe = Vec3::new(lf_z, lf_y, -lf_x);
    let len = local_globe.length();
    if len < 1e-12 {
        Vec3::ZERO
    } else {
        local_globe.scale(1.0 / len)
    }
}

pub fn ra_dec_to_az_el(
    ra_rad: f64,
    dec_rad: f64,
    lat_deg: f64,
    lon_deg: f64,
    gmst_deg: f64,
) -> Angles {
    if !ra_rad.is_finite() || !dec_rad.is_finite() {
        return Angles {
            azimuth: f64::NAN,
            elevation: f64::NAN,
        };
    }
    let lat = to_rad(lat_deg);
    let lst = to_rad(gmst_deg + lon_deg);
    let ha = lst - ra_rad;
    let sin_alt = lat.sin() * dec_rad.sin() + lat.cos() * dec_rad.cos() * ha.cos();
    let alt = limit1(sin_alt).asin();
    let y = -dec_rad.cos() * ha.sin();
    let x = dec_rad.sin() * lat.cos() - dec_rad.cos() * lat.sin() * ha.cos();
    let mut az = to_deg(y.atan2(x));
    az = ((az % 360.0) + 360.0) % 360.0;
    Angles {
        azimuth: az,
        elevation: to_deg(alt),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn local_globe_angles_follow_axis_convention() {
        let zenith = local_globe_coord_to_angles(Vec3::new(1.0, 0.0, 0.0));
        assert_eq!(zenith.azimuth, 0.0);
        assert!((zenith.elevation - 90.0).abs() < 1e-12);

        let east_horizon = local_globe_coord_to_angles(Vec3::new(0.0, 1.0, 0.0));
        assert!((east_horizon.azimuth - 90.0).abs() < 1e-12);
        assert!(east_horizon.elevation.abs() < 1e-12);
    }
}
